chips/msp432/src/dma.rs

```diff
@@ -0,0 +1,916 @@
+
+        // Set the the number of cycles after the module rearbitrates the bus.
+        // The 'register-value' for this is ld(cycles), e.g. cycles = 256 -> ld(256) = 8
+        // In order to get the number of cycles, just get the closest 2^n value of transfers
+        let r_power = if transfers > MAX_TRANSFERS_LEN {
+            31 - (MAX_TRANSFERS_LEN as u32).leading_zeros()
+        } else {
+            31 - (len as u32).leading_zeros()
+        };
+
+        // Set the number of cycles before a bus rearbitration
+        DMA_CONFIG.0[self.chan_nr]
+            .ctrl
+            .modify(DMA_CTRL::R_POWER.val(r_power));
+
+        // Store the buffers
+        self.rx_buf.replace(dst_buf);
+        self.tx_buf.replace(src_buf);
+
+        // Store transfer-type
+        self.transfer_type.set(DmaTransferType::MemoryToMemory);
+
+        // Enable the DMA channel
+        self.enable_dma_channel();
+    }
+
+    /// Start a DMA transfer where the contents of any register will be copied into a provided buffer
+    pub fn transfer_periph_to_mem(&self, src_reg: *const (), buf: &'static mut [u8], len: usize) {
+        // The pointers must point to the end of the buffer, for detailed calculation see
+        // datasheet p. 646, section 11.2.4.4.
+        let src_end_ptr = src_reg as u32;
+        let dst_end_ptr = (&buf[0] as *const u8 as u32) + ((len as u32) - 1);
+
+        self.setup_transfer(src_end_ptr, dst_end_ptr, len);
+
+        // Store the buffer
+        self.rx_buf.replace(buf);
+
+        // Store transfer-type
+        self.transfer_type.set(DmaTransferType::PeripheralToMemory);
+
+        self.enable_dma_channel();
+    }
+
+    /// Start a DMA transfer where the contents of a buffer will be copied into a register
+    pub fn transfer_mem_to_periph(&self, dst_reg: *const (), buf: &'static mut [u8], len: usize) {
+        // The pointers must point to the end of the buffer, for detailed calculation see
+        // datasheet p. 646, section 11.2.4.4.
+        let src_end_ptr = (&buf[0] as *const u8 as u32) + ((len as u32) - 1);
+        let dst_end_ptr = dst_reg as u32;
+
+        self.setup_transfer(src_end_ptr, dst_end_ptr, len);
+
+        // Store the buffer
+        self.tx_buf.replace(buf);
+
+        // Store transfer-type
+        self.transfer_type.set(DmaTransferType::MemoryToPeripheral);
+
+        self.enable_dma_channel();
+    }
+}
+
+pub fn handle_interrupt(int_nr: isize) {
+    if int_nr == 0 {
+        // For now only use the INT0 because I don't know how to prioritize the channels in order
+        // to give them 1 out of 3 'own' interrupts.
+        let int = DMA_BASE.int0_srcflg.get();
+
+        for i in 0..AVAILABLE_DMA_CHANNELS {
+            let bit = (1 << i) as u32;
+            if (bit & int) > 0 {
+                unsafe {
+                    // Clear interrupt-bit
+                    DMA_BASE.int0_clrflg.set(bit);
+
+                    DMA_CHANNELS[i].handle_interrupt();
+                }
```

Comment:
Can you add a comment here explaining why this unsafe block is needed?

Also, I think you can move the `set()` part out of the unsafe block, no?

Reply:
Yes you are right, `set()` can be moved out, thanks for the hint!

I didn't add a comment to the `unsafe` block because it is accessing a global mutable variable, so it seemed obvious to me that this must be unsafe. But I will add a comment here (: